File: Review/ShellSort.cpp
#include <iostream>

using namespace std;

template <typename T>
void shellSort(T arr[], int n){

	int step = n;
	for(step >> 1; step > 0; step >> 1){

		int j;
		for(int i = step - 1; i < n; i ++){

			T e = arr[i];
			for(j = i; j > step - 1 && arr[j - step] > e; j -= step)
				arr[j] = arr[j - step];
			arr[j] = e;
		}
	}

	return;
}



#include <iostream>

using namespace std;

template <typename T>
void shellSort(T arr[], int n){

	int step = n;
	for(step >>= 1; step > 0; step >>= 1){

		for(i = step - 1; i < n; i ++){

			T e = arr[i];
			int j;
			for(j = i; j > step - 1 && arr[j - step] > e; j -= step)
				arr[j] = arr[j - step];
			arr[j] = e;
		}
	}

	return;
}

File: Review/ShellSort.py
def shellSort(arr, n):
	step = n / 2
	while step > 0:
		for i in range(step - 1, n):
			e = arr[i]
			for j in range(i, step - 1, -step):
				if arr[j - step] > e:
					arr[j] = arr[j - step]
				else:
					break
			arr[j] = e
		step = step / 2

	return


def shellSort(arr, n):
	step = n / 2
	while step > 0:
		for i in range(step - 1, n):
			e = arr[i]
			j = i
			while j > step - 1 and arr[j - step] > e:
				arr[j] = arr[j - step]
				j -= step
			arr[j] = e
		step = step / 2

	return arr